### PR TITLE
client: Support parallel TSO RPC requests on single dispatcher loop

### LINKED ISSUE #8432: client: Support TSO RPC Parallelizing
## Development Task

- The main tracking issue in TiDB repo: https://github.com/pingcap/tidb/issues/54960

We used to find that in some OLTP workloads where the QPS is high and the queries are simple, the TSO Wait duration usually become a significant portion of the total duration of queries. In TiDB, TSO loading is already made concurrent with some other works such as compiling. In cases that the queries are simple, it would be hard to further optimize it by making it concurrent with more phases of the SQL execution. But we found a practical way to optimize it is to do it from the TSO client.

Currently, a TSO client object has a goroutine that collects `GetTS` (and `GetTSAsync`) calls (`tsoRequest`s) as a batch, send it to PD, wait for the response, and dispatch the results to these `tsoRequests`, serially. As a result, each `GetTS` calls may need to spend up to 1x TSO RPC time to wait for being collected to the next batch.

Considering the case that PD's TSO allocator is not the bottle neck and can deal with more TSO requests (so that the majority part of TSO RPC's time cost is on the network), we find that it's possible to start collecting the next batch and send it before receiving the response of the previous batch. So that each `GetTS` call needs to wait for less time to be batched, and gets a shorter total duration.

So this is an approach that reduces the duration of `GetTS` & `GetTSAsync - Wait` at the expense of higher TSO RPC OPS and higher pressure to PD. It's not suitable to be enabled by default, but we can provide such an option when the TSO Wait duration becomes a problem.

### Subtasks

Side changes:

- [x] Merge the two `xxxTSOStream` types  so that the error handling and metrics reporting logic for PD server deployment and TSO service deployment can be reused. https://github.com/tikv/pd/pull/8433
- [x] Split sending and receiving part of `tsoStream` into separated goroutines
- [ ] Let `tsoDispatcher` support batching according to estimated TSO RPC duration from `tsoStream`

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What problem does this PR solve?

Issue Number: Close #8432

### What is changed and how does it work?

```commit-message
client: Support parallel TSO RPC requests on single dispatcher loop

This commit supports handling multiple TSO RPC concurrently in one single dispatcher loop to reduce the expected time that each GetTS call spent on waiting the next batch.
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/tso_dispatcher.go [hunk 1]
```diff
@@ -578,5 +650,98 @@ func (td *tsoDispatcher) compareAndSwapTS(
...
+	// because that MaxTSOBatchWaitInterval is just enabled. In this case, disable concurrent TSO RPC requests
+	// immediately, because MaxTSOBatchWaitInterval and concurrent RPC requests has opposite purpose.
+	immediatelyUpdate := td.rpcConcurrency > 1 && maxBatchWaitInterval > 0
+
+	// Allow always updating for test purpose.
+	failpoint.Inject("tsoDispatcherAlwaysCheckConcurrency", func() {
+		immediatelyUpdate = true
+	})
+
+	if !immediatelyUpdate && now.Sub(td.lastCheckConcurrencyTime) < dispatcherCheckRPCConcurrencyInterval {
```
Comment (JmPotato) [inline client/tso_dispatcher.go:698]: Don't we need to update `td.lastCheckConcurrencyTime` here also?

File: client/tso_stream.go [hunk 2]
```diff
@@ -228,6 +232,8 @@ var streamIDAlloc atomic.Int32
 
 const invalidStreamID = "<invalid>"
 
+const maxPendingRequestsInTSOStream = 64
```
Comment (JmPotato) [inline client/tso_stream.go:235]: Please merge with line 233.

File: client/tso_dispatcher.go [hunk 3]
```diff
@@ -542,32 +619,27 @@ func (td *tsoDispatcher) doneCollectedRequests(tbc *tsoBatchController, physical
...
 				zap.Uint32("new-group-id", curTSOInfo.respKeyspaceGroupID))
 		}
 		// The TSO we get is a range like [largestLogical-count+1, largestLogical], so we save the last TSO's largest logical
 		// to compare with the new TSO's first logical. For example, if we have a TSO resp with logical 10, count 5, then
-		// all TSOs we get will be [6, 7, 8, 9, 10]. lastTSOInfo.logical stores the logical part of the largest ts returned
+		// all TSOs we get will be [6, 7, 8, 9, 10]. latestTSOInfo.logical stores the logical part of the largest ts returned
 		// last time.
-		if tsoutil.TSLessEqual(physical, firstLogical, lastTSOInfo.physical, lastTSOInfo.logical) {
+		if tsoutil.TSLessEqual(curTSOInfo.physical, firstLogical, lastTSOInfo.physical, lastTSOInfo.logical) {
 			log.Panic("[tso] timestamp fallback",
```
Comment (JmPotato) [inline client/tso_dispatcher.go:638]: Shall we print out the stream ID here?

File: client/tso_dispatcher.go (hunk 1 above)
Comment (MyonKeminta) [reply client/tso_dispatcher.go:698]: It's updated at line 701. We don't update it if it's skipped, otherwise it may never be checked

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**